src/detect-flowbits.c

```diff
@@ -745,6 +860,482 @@ static void DetectFlowbitsAnalyzeDump(const DetectEngineCtx *de_ctx,
+        if (add->rule_id == NULL) {
+            SCFree(add);
+            return -1;
+        }
+
+        if (!AddSids(de_ctx, fba, fb, add)) {
+            SCLogDebug("no sids added");
+            SCFree(add->rule_id);
+            SCFree(add);
+            return 0;
+        }
+        PrefilterFlowbit *res = PFB_RB_INSERT(&ctx->fb_tree, add);
+        SCLogDebug("not found, so added (res %p)", res);
+        BUG_ON(res != NULL); // TODO if res != NULL we have a duplicate which should be impossible
+    } else {
+        SCLogDebug("found! pfb %p id %u", pfb, pfb->id);
+
+        int r = AddSids(de_ctx, fba, fb, pfb);
+        if (r < 0) {
+            return -1;
+        } else if (r == 0) {
+            SCLogDebug("no sids added");
+            return 0;
+        }
+    }
+    return 1;
+}
+
+/** \brief build flowbit prefilter state(s)
+ *
+ *  Build "set" and "isset" states.
+ *
+ *  For each flowbit "isset" in the sgh, we need to check:
+ *  1. is it supported
+ *  2. is prefilter enabled
+ *  3. does it match in the same dir or only opposing dir
+ */
+static int PrefilterSetupFlowbits(DetectEngineCtx *de_ctx, SigGroupHead *sgh)
+{
+    if (sgh == NULL)
+        return 0;
+
+    SCLogDebug("sgh %p: setting up prefilter", sgh);
+    struct FBAnalyzer fb_analysis = DetectFlowbitsAnalyzeForGroup(de_ctx, sgh);
+    struct PrefilterEngineFlowbits *isset_ctx = NULL;
+    struct PrefilterEngineFlowbits *set_ctx = NULL;
+
+    for (uint32_t i = 0; i < sgh->init->sig_cnt; i++) {
+        Signature *s = sgh->init->match_array[i];
+        if (s == NULL)
+            continue;
+
+        SCLogDebug("checking sid %u", s->id);
+
+        /* first build the 'set' state */
+        for (SigMatch *sm = s->init_data->smlists[DETECT_SM_LIST_POSTMATCH]; sm != NULL;
+                sm = sm->next) {
+            if (sm->type != DETECT_FLOWBITS) {
+                SCLogDebug("skip non flowbits sm");
+                continue;
+            }
+
+            DetectFlowbitsData *fb = (DetectFlowbitsData *)sm->ctx;
+            if (fb->cmd == DETECT_FLOWBITS_CMD_SET) {
+                SCLogDebug(
+                        "DETECT_SM_LIST_POSTMATCH: sid %u DETECT_FLOWBITS set %u", s->id, fb->idx);
+                // else if (fb->cmd == DETECT_FLOWBITS_CMD_TOGGLE) {
+                //   SCLogDebug("DETECT_SM_LIST_POSTMATCH: sid %u DETECT_FLOWBITS toggle %u", s->id,
+                //           fb->idx);
+            } else {
+                SCLogDebug("unsupported flowbits setting");
+                continue;
+            }
+
+            if (fb_analysis.array[fb->idx].toggle_sids_idx ||
+                    fb_analysis.array[fb->idx].isnotset_sids_idx ||
+                    fb_analysis.array[fb->idx].unset_sids_idx) {
+                SCLogDebug("flowbit %u not supported: toggle or unset in use", fb->idx);
+                continue;
+            }
+
+            if (set_ctx == NULL) {
+                set_ctx = SCCalloc(1, sizeof(*set_ctx));
```

Comment:
this "set" post-rule-prefilter state should only be built if there are isset flowbit rules, with prefilter enabled, in the same sgh. Need to review is this is checked correctly.